Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp

```diff
@@ -0,0 +1,244 @@
+}
+void PolygonalSelectTool::on_mousedown(Layer*, MouseEvent& event)
+{
+    auto& image_event = event.image_event();
+    if (image_event.button() != GUI::MouseButton::Primary)
+        return;
+    if (!m_selecting) {
+        m_polygon_points.clear();
+        m_last_selecting_cursor_position = event.layer_event().position();
+    }
+
+    m_selecting = true;
+
+    auto new_point = event.layer_event().position();
+
+    // this point matches the first point exactly. consider this polygon finished
+    if (m_polygon_points.size() > 0 && new_point == m_polygon_points.at(0)) {
+        m_selecting = false;
+        m_editor->image().selection().end_interactive_selection();
+        process_polygon();
+        m_editor->did_complete_action(tool_name());
+        m_editor->update();
+        return;
+    }
+
+    // prevent duplicate points back to back
+    if (m_polygon_points.size() > 0 && m_polygon_points.at(m_polygon_points.size() - 1) == new_point)
+        return;
+
+    m_polygon_points.append(new_point);
+    m_editor->image().selection().begin_interactive_selection();
+
+    m_editor->update();
+}
+
+void PolygonalSelectTool::on_mousemove(Layer*, MouseEvent& event)
+{
+    if (m_selecting)
+        m_last_selecting_cursor_position = event.layer_event().position();
+    m_editor->update();
+}
+void PolygonalSelectTool::on_doubleclick(Layer*, MouseEvent&)
+{
+    m_selecting = false;
+    m_editor->image().selection().end_interactive_selection();
+    process_polygon();
+    m_editor->did_complete_action(tool_name());
+    m_editor->update();
+}
+void PolygonalSelectTool::on_second_paint(Layer const* layer, GUI::PaintEvent& event)
+{
+    if (!m_selecting)
+        return;
+
+    GUI::Painter painter(*m_editor);
+    painter.add_clip_rect(event.rect());
+
+    painter.translate(editor_layer_location(*layer));
+
+    for (size_t i = 0; i < m_polygon_points.size() - 1; i++) {
+        auto preview_start = editor_stroke_position(m_polygon_points.at(i), 1);
+        auto preview_end = editor_stroke_position(m_polygon_points.at(i + 1), 1);
+        painter.draw_line(preview_start, preview_end, Color::Black, AK::max(m_editor->scale(), 1));
+    }
+
+    auto last_line_start = editor_stroke_position(m_polygon_points.at(m_polygon_points.size() - 1), 1);
+    auto last_line_stop = editor_stroke_position(m_last_selecting_cursor_position, 1);
+    painter.draw_line(last_line_start, last_line_stop, Color::Black, AK::max(m_editor->scale(), 1));
+}
+
+void PolygonalSelectTool::on_keydown(GUI::KeyEvent& key_event)
+{
+    Tool::on_keydown(key_event);
+    if (key_event.key() == KeyCode::Key_Escape) {
+        if (m_selecting) {
+            m_selecting = false;
+            m_polygon_points.clear();
+        } else
+            m_editor->image().selection().clear();
+    }
+}
+
+GUI::Widget* PolygonalSelectTool::get_properties_widget()
+{
+    if (m_properties_widget) {
+        return m_properties_widget.ptr();
+    }
```

Comment:
Tiny nit: You can remove the braces here

Reply:
Fixed!